### ISSUE TITLE
Feature request: multithreaded `pheweb matrix`

### ISSUE THREAD

**Issue by jgrundstad** [statgen/pheweb#115]
Would it be possible to perform the matrix processing stage on a per-chromosome basis to allow distribution of the work?

Thanks!

**Comment by pjvandehaar** [comment]
1. There's two ways to deal with the single-threaded time taken by `pheweb matrix`:

(a) Parallelize `pheweb matrix`.

(b) Instead of running `pheweb process`, run

```
    pheweb phenolist verify &&
    pheweb parse-input-files &&
    pheweb sites &&
    pheweb make-gene-aliases-trie &&
    pheweb add-rsids &&
    pheweb add-genes &&
    pheweb make-tries &&
    pheweb augment-phenos
```

and then, in parallel, run both `pheweb matrix` and

```
    pheweb manhattan &&
    pheweb qq &&
    pheweb bgzip-phenos &&
    pheweb top-hits &&
    pheweb phenotypes &&
    pheweb gather-pvalues-for-each-gene
```

If the single threaded `pheweb matrix` finishes first, this solution is likely faster.

I'm not sure how to implement this option well.  If pheweb used `multiprocessing` to run `pheweb matrix` in parallel with the other commands, debugging and terminal output will be harder.  Perhaps I could add an option `pheweb process --matrix-in-parallel`, in which `pheweb matrix` output would be redirected to a log file?

Simply because of the debugging complexity, I like option (a), like you proposed.

---

2. Currently, pheweb makes the matrix from the un-compressed per-phenotype annotated files.  (`pheno/*` on https://github.com/statgen/pheweb/blob/master/etc/detailed-internal-dataflow.md).  Since these are uncompressed, they're not indexed.

To read a single chromosome from an individual phenotype file for use in `pheweb matrix`, either:

* Each reader does a bisect to find the line to start reading on
* Read directly from the tabixed files (which might be slow)

---

3. Are you interested in trying to implement this?

**Comment by jgrundstad** [comment]
I'm not sure what the best route is, still getting a feeling for how this whole process works :)   Unfortunately bandwidth limits will keep me in a user role for the time being.

I'm willing to try option 1.b. in the near future - I didn't realize stages could be run concurrently.

I'm currently processing a set of 12m snps, 2000 phenotypes, early calculations indicate about 100hours of processing for the `pheweb matrix` stage.  It's running one core at ~35% load and should be done in 3.5 days.  How long would you anticipate the remaining steps should take?

**Comment by pjvandehaar** [comment]
How many cores on the machine?

`pheweb process -h` will show all the steps it runs.  The only heavy one remaining is `pheweb gather-pvalues-for-each-gene`, which might take as long as `pheweb matrix` with 15 cores.  Once it starts, it should give a good estimate.

Do you predict runtime with something like `gzip -cd .generated-by-pheweb/tmp/matrix.tsv.gz | wc -l`?

**Comment by hanayik** [comment]
Hi, does anyone in this thread know of a branch or fork that has implemented a parallelized `pheweb matrix`?

**Comment by pjvandehaar** [comment]
@hanayik I'm not aware of such a fork.  I'd love to have a parallel matrix feature and I think it shouldn't be too hard if you'd like to work on it.  Changes would be needed in https://github.com/statgen/pheweb/blob/master/pheweb/load/matrix.py and https://github.com/statgen/pheweb/blob/master/pheweb/load/cffi/x.cpp .

Some considerations:
1. Parallelization will have to happen by genomic region, but which regions should be used?  Should each process handle one chromosome, or one 1-megabase block (in which case we'd need to handle indels spanning block boundaries)?
2. Each region's partial matrix tsv must be written already gzipped to avoid filling the disk.  To avoid un-gzipping and re-gzipping when concatenating the regional partial matrix tsv files, we'll need to handle the bgzip empty tail block (either by deleting all but the last, or by not writing them in the first place).
3. Should the forking happen in C++ (pthreads?) or Python (`multiprocessing`)?

**Comment by Shicheng-Guo** [comment]
Looks not so bad. I have 1262 phenotypes and, with single thread, `pheweb matrix` only takes 62880 seconds. I have a question: Is there any way to speed up `pheweb bgzip-phenos` step?

```
==> Starting `pheweb phenolist verify`
The 1262 phenotypes in '/projects/ps-janssen4/dsci-csb/user/sguo2/pheweb/pheno-list.json' look good.
==> Completed in 0 seconds

==> Starting `pheweb sites`
The list of sites is up-to-date!
==> Completed in 1 seconds

==> Starting `pheweb make-gene-aliases-trie`
gene aliases are at '/home/sguo2/.pheweb/cache/gene_aliases-v29-hg19.marisa_trie'
==> Completed in 0 seconds

==> Starting `pheweb add-rsids`
rsid annotation is up-to-date!
==> Completed in 0 seconds

==> Starting `pheweb add-genes`
gene annotation is up-to-date!
==> Completed in 0 seconds

==> Starting `pheweb make-tries`
tries are up-to-date!
==> Completed in 0 seconds

==> Starting `pheweb augment-phenos`
Output files are all newer than input files, so there's nothing to do.
==> Completed in 1 seconds

==> Starting `pheweb manhattan`
Output files are all newer than input files, so there's nothing to do.
==> Completed in 1 seconds

==> Starting `pheweb qq`
Output files are all newer than input files, so there's nothing to do.
==> Completed in 2 seconds

==> Starting `pheweb matrix`
==> Completed in 62880 seconds

==> Starting `pheweb bgzip-phenos`
Processing 1262 phenos
```

**Comment by pjvandehaar** [comment]
@Shicheng-Guo Yeah, if your system is CPU-limited (rather than IO-limited), then reducing the gzip compression level will probably make `pheweb bgzip-phenos` a 2x-3x faster.  Unfortunately, I don't think pysam lets you choose gzip compression level.  It looks easy to switch from `pysam.tabix_compress` to the code in `load/cffi/x.cpp`, but I personally am not going to do it, because I only run pheweb on machines with plenty of CPU.  If you'd like to work on this, I'm happy to give advice.

**Comment by Shicheng-Guo** [comment]
Thanks Peter. I think I have enough CPU.  For me, it would be perfect If I can submit 1262 jobs to run all these 1262 bgzip-phenos simultaneously. I am wondering whether the following script work or not? 

```
#!/bin/bash
#SBATCH --cpus-per-task=10
#SBATCH --mem=20G
#SBATCH --time=48:50:00
#SBATCH --array=0-126
jobs=(0..1264)
export PHEWEB_DATADIR='/projects/ps-janssen4/dsci-csb/user/sguo2/pheweb'
/home/sguo2/janssen4/bin/miniconda3/bin/pheweb conf num_procs=10 parse --phenos=${jobs[$SLURM_ARRAY_TASK_ID]}

```

Here is my current situation which is not so good. I used `pheweb conf num_procs=50 process --no-parse`, but it looks all the 50 jobs running in same node and then each of them only take 2% CPU which make the processing very slow. 

![image](https://user-images.githubusercontent.com/6641283/98334167-47037680-1fd0-11eb-80d3-335b4674d7a8.png)

Shicheng

**Comment by pjvandehaar** [comment]
Could you move this to a new issue?
